File: src-tauri/src/domain/account.rs
use serde::{Deserialize, Serialize};

use crate::domain::provider::ProviderKind;
use crate::domain::types::AccountId;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionVerificationStatus {
    Verified,
    #[default]
    Unverified,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Account {
    pub id: AccountId,
    pub kind: ProviderKind,
    pub name: String,
    pub server_url: Option<String>,
    pub username: Option<String>,
    pub sync_interval_secs: i64,
    pub sync_on_startup: bool,
    pub sync_on_wake: bool,
    pub keep_read_items_days: i64,
    pub connection_verification_status: ConnectionVerificationStatus,
    pub connection_verified_at: Option<String>,
    pub connection_verification_error: Option<String>,
}

File: src/api/schemas/account.ts
import * as v from "valibot";
import * as s from "@/api/schemas/validation";
import { IsoDateTimeStringSchema } from "./common";

const AccountConnectionVerificationStatusSchema = v.picklist(["verified", "unverified", "error", "quarantined"]);
const accountSyncIntervalSecsSchema = v.pipe(v.number(), v.integer(), v.finite(), v.minValue(60), v.maxValue(86_400));
const accountKeepReadItemsDaysSchema = v.pipe(v.number(), v.integer(), v.finite(), v.minValue(0), v.maxValue(3650));
const nonBlankTrimmedStringSchema = v.pipe(v.string(), v.trim(), v.minLength(1));
const optionalBlankStringToUndefinedSchema = v.optional(
  v.pipe(
    v.unknown(),
    v.transform((value) => {
      if (typeof value !== "string") {
        return value;
      }
      const trimmed = value.trim();
      return trimmed.length > 0 ? trimmed : undefined;
    }),
    v.optional(v.string()),
  ),
);

const AccountProviderCapabilitiesSchema = s.strictObject({
  supports_folders: v.boolean(),
  supports_starring: v.boolean(),
  supports_search: v.boolean(),
  supports_delta_sync: v.boolean(),
  supports_remote_state: v.boolean(),
});

export const AccountDtoSchema = s.strictObject({
  id: nonBlankTrimmedStringSchema,
  kind: nonBlankTrimmedStringSchema,
  name: nonBlankTrimmedStringSchema,
  display_name: optionalBlankStringToUndefinedSchema,
  icon_url: v.optional(v.nullable(v.string())),
  capabilities: v.optional(AccountProviderCapabilitiesSchema),
  server_url: v.nullable(v.string()),
  username: v.nullable(v.string()),
  sync_interval_secs: accountSyncIntervalSecsSchema,
  sync_on_startup: v.boolean(),
  sync_on_wake: v.boolean(),
  keep_read_items_days: accountKeepReadItemsDaysSchema,
  connection_verification_status: v.optional(AccountConnectionVerificationStatusSchema),
  connection_verified_at: v.optional(v.nullable(IsoDateTimeStringSchema)),
  connection_verification_error: v.optional(v.nullable(v.string())),
});

export const AccountDtoListSchema = v.array(AccountDtoSchema);

export type AccountDto = v.InferOutput<typeof AccountDtoSchema>;
